[apr] Build error with cross compile

Package: apr:arm64-windows -> 1.7.5

**Host Environment**

- Host: x64-windows
- Compiler: MSVC 19.39.33521.0
-    vcpkg-tool version: 2023-08-09-9990a4c9026811a312cb2af78bf77f3d9d288416
    vcpkg-scripts version: 91d888703 2024-09-04 (11 hours ago)

**To Reproduce**

`vcpkg install apr:arm64-windows`
**Failure logs**

```
-- Downloading https://archive.apache.org/dist/apr/apr-1.7.5.tar.bz2 -> apr-1.7.5.tar.bz2...
-- Extracting source F:/git/vcpkg-x64/downloads/apr-1.7.5.tar.bz2
-- Applying patch unglue.patch
-- Using source at F:/git/vcpkg-x64/buildtrees/apr/src/apr-1-bc774307b8.clean
-- Configuring arm64-windows
-- Building arm64-windows-rel
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: F:/git/vcpkg-x64/downloads/tools/cmake-3.29.2-windows/cmake-3.29.2-windows-i386/bin/cmake.exe --build . --config Release --target install -- -v -j17
    Working Directory: F:/git/vcpkg-x64/buildtrees/apr/arm64-windows-rel
    See logs for more information:
      F:\git\vcpkg-x64\buildtrees\apr\install-arm64-windows-rel-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/apr/portfile.cmake:37 (vcpkg_cmake_install)
  scripts/ports.cmake:192 (include)



```
<details><summary>F:\git\vcpkg-x64\buildtrees\apr\install-arm64-windows-rel-out.log</summary>

```
Change Dir: 'F:/git/vcpkg-x64/buildtrees/apr/arm64-windows-rel'

Run Build Command(s): F:/git/vcpkg-x64/downloads/tools/ninja/1.10.2-windows/ninja.exe -v -v -j17 install
[1/85] C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\arm64\cl.exe   -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -IF:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\arch\win32 -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\arch\unix -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\private /nologo /DWIN32 /D_WINDOWS /utf-8 /MP  /MD /O2 /Oi /Gy /DNDEBUG /Z7 /showIncludes /FoCMakeFiles\gen_test_char.dir\tools\gen_test_char.c.obj /FdCMakeFiles\gen_test_char.dir\ /FS -c F:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\tools\gen_test_char.c
[2/85] C:\WINDOWS\system32\cmd.exe /C "cd . && F:\git\vcpkg-x64\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=CMakeFiles\gen_test_char.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\arm64\link.exe  CMakeFiles\gen_test_char.dir\tools\gen_test_char.c.obj  /out:gen_test_char.exe /implib:gen_test_char.lib /pdb:gen_test_char.pdb /version:0.0 /machine:ARM64 /nologo /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF   /subsystem:console  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
[3/85] C:\WINDOWS\system32\cmd.exe /C "cd /D F:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel && F:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel\gen_test_char.exe > F:/git/vcpkg-x64/buildtrees/apr/arm64-windows-rel/apr_escape_test_char.h"
FAILED: apr_escape_test_char.h F:/git/vcpkg-x64/buildtrees/apr/arm64-windows-rel/apr_escape_test_char.h 
C:\WINDOWS\system32\cmd.exe /C "cd /D F:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel && F:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel\gen_test_char.exe > F:/git/vcpkg-x64/buildtrees/apr/arm64-windows-rel/apr_escape_test_char.h"
This version of F:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel\gen_test_char.exe is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher.
[4/85] C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\arm64\cl.exe   -DAPR_APP -DWINNT -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -IF:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\arch\win32 -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\arch\unix -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\private /nologo /DWIN32 /D_WINDOWS /utf-8 /MP  /MD /O2 /Oi /Gy /DNDEBUG /Z7 /showIncludes /FoCMakeFiles\libaprapp-1.dir\misc\win32\apr_app.c.obj /FdCMakeFiles\libaprapp-1.dir\libaprapp-1.pdb /FS -c F:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\misc\win32\apr_app.c
[5/85] C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\arm64\cl.exe   -DAPR_APP -DWINNT -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -IF:\git\vcpkg-x64\buildtrees\apr\arm64-windows-rel -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\arch\win32 -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\arch\unix -IF:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\include\private /nologo /DWIN32 /D_WINDOWS /utf-8 /MP  /MD /O2 /Oi /Gy /DNDEBUG /Z7 /showIncludes /FoCMakeFiles\libaprapp-1.dir\misc\win32\internal.c.obj /FdCMakeFiles\libaprapp-1.dir\libaprapp-1.pdb /FS -c F:\git\vcpkg-x64\buildtrees\apr\src\apr-1-bc774307b8.clean\misc\win32\internal.c
ninja: build stopped: subcommand failed.
```
</details>

--- a/ports/activemq-cpp/vcpkg.json
+++ b/ports/activemq-cpp/vcpkg.json
@@ -1,10 +1,10 @@
 {
   "name": "activemq-cpp",
   "version-semver": "3.9.5",
-  "port-version": 16,
+  "port-version": 17,
   "description": "Apache ActiveMQ is the most popular and powerful open source messaging and Integration Patterns server.",
   "license": "Apache-2.0",
-  "supports": "!(uwp | osx)",
+  "supports": "(windows & !uwp & (x86 | x64)) | (!windows & !osx)",
   "dependencies": [
     "apr",
     {

--- a/ports/apr/0100-add-host-tools-dir.diff
+++ b/ports/apr/0100-add-host-tools-dir.diff
@@ -1,0 +1,23 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index d16eec6..92146f4 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -87,13 +87,17 @@ STRING(REGEX REPLACE ".*#define APR_PATCH_VERSION[ \t]+([0-9]+).*" "\\1" APR_PAT
+ 
+ CONFIGURE_FILE(include/apr.hwc
+                ${PROJECT_BINARY_DIR}/apr.h)
+
+ ADD_EXECUTABLE(gen_test_char tools/gen_test_char.c)
+
++set(UNOFFICIAL_APR_HOST_TOOLS_DIR "$<TARGET_FILE_DIR:gen_test_char>" CACHE STRING "")
++set(UNOFFICIAL_APR_HOST_EXECUTABLE_SUFFIX "$<TARGET_PROPERTY:gen_test_char,SUFFIX>" CACHE STRING "")
++install(TARGETS gen_test_char)
++
+ ADD_CUSTOM_COMMAND(
+   COMMENT "Generating character tables, apr_escape_test_char.h, for current locale"
+   DEPENDS gen_test_char
+-  COMMAND $<TARGET_FILE:gen_test_char> > ${PROJECT_BINARY_DIR}/apr_escape_test_char.h
++  COMMAND "${UNOFFICIAL_APR_HOST_TOOLS_DIR}/gen_test_char${UNOFFICIAL_APR_HOST_EXECUTABLE_SUFFIX}" > ${PROJECT_BINARY_DIR}/apr_escape_test_char.h
+   OUTPUT ${PROJECT_BINARY_DIR}/apr_escape_test_char.h
+ )
+ ADD_CUSTOM_TARGET(

--- a/ports/apr/portfile.cmake
+++ b/ports/apr/portfile.cmake
@@ -1,6 +1,3 @@
-
-set(VERSION 1.7.5)
-
 vcpkg_download_distfile(ARCHIVE
     URLS "https://archive.apache.org/dist/apr/apr-${VERSION}.tar.bz2"
     FILENAME "apr-${VERSION}.tar.bz2"
@@ -11,7 +8,18 @@ vcpkg_extract_source_archive(SOURCE_PATH
     ARCHIVE "${ARCHIVE}"
     PATCHES
         unglue.patch
+        0100-add-host-tools-dir.diff
 )
+
+set(CURRENT_HOST_TOOLS_DIR "${CURRENT_HOST_INSTALLED_DIR}/tools/${PORT}")
+
+set(CROSSCOMPILING_OPTIONS "")
+if(VCPKG_CROSSCOMPILING)
+    list(APPEND CROSSCOMPILING_OPTIONS
+        "-DUNOFFICIAL_APR_HOST_TOOLS_DIR=${CURRENT_HOST_TOOLS_DIR}"
+        "-DUNOFFICIAL_APR_HOST_EXECUTABLE_SUFFIX=${VCPKG_HOST_EXECUTABLE_SUFFIX}"
+    )
+endif()
 
 if (VCPKG_TARGET_IS_WINDOWS)
     vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
@@ -32,6 +40,7 @@ if (VCPKG_TARGET_IS_WINDOWS)
             -DMIN_WINDOWS_VER=Windows7
             -DAPR_HAVE_IPV6=ON
             ${FEATURE_OPTIONS}
+            ${CROSSCOMPILING_OPTIONS}
     )
 
     vcpkg_cmake_install()
@@ -42,6 +51,7 @@ if (VCPKG_TARGET_IS_WINDOWS)
     )
     # There is no way to suppress installation of the headers in debug builds.
     file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+    vcpkg_copy_tools(TOOL_NAMES gen_test_char AUTO_CLEAN)
 
     vcpkg_copy_pdbs()
 

--- a/ports/apr/vcpkg.json
+++ b/ports/apr/vcpkg.json
@@ -1,12 +1,17 @@
 {
   "name": "apr",
   "version": "1.7.5",
-  "port-version": 1,
+  "port-version": 2,
   "description": "The Apache Portable Runtime (APR) is a C library that forms a system portability layer that covers many operating systems.",
   "homepage": "https://apr.apache.org/",
   "license": "Apache-2.0",
-  "supports": "!uwp",
+  "supports": "!uwp & !mingw",
   "dependencies": [
+    {
+      "name": "apr",
+      "host": true,
+      "platform": "windows"
+    },
     {
       "name": "vcpkg-cmake",
       "host": true,

--- a/scripts/ci.baseline.txt
+++ b/scripts/ci.baseline.txt
@@ -51,8 +51,6 @@ allegro5:x64-android=fail
 apr:arm-neon-android=fail
 apr:arm64-android=fail
 apr:x64-android=fail
-# Cross compiling CI machine cannot run gen_test_char to generate apr_escape_test_char.h
-apr:arm64-windows=fail
 apsi:arm-neon-android=fail
 apsi:x64-android=fail
 # Broken with CUDA 12; needs update to 3.8.3 and https://github.com/arrayfire/arrayfire/issues/3349 fixed

--- a/versions/a-/activemq-cpp.json
+++ b/versions/a-/activemq-cpp.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "0066f657df214848cbfd6b991e8ff4a52902e81b",
+      "version-semver": "3.9.5",
+      "port-version": 17
+    },
+    {
       "git-tree": "1644c99e45b2364ea36c0307b4d7171f4717ccdc",
       "version-semver": "3.9.5",
       "port-version": 16

--- a/versions/a-/apr.json
+++ b/versions/a-/apr.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "dfa11d3e5058b640b37c1a6845dbc2980496a7c8",
+      "version": "1.7.5",
+      "port-version": 2
+    },
+    {
       "git-tree": "499cf5fc0959f4a049ec72f0f556400e7191ffd3",
       "version": "1.7.5",
       "port-version": 1

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -34,7 +34,7 @@
     },
     "activemq-cpp": {
       "baseline": "3.9.5",
-      "port-version": 16
+      "port-version": 17
     },
     "ada-url": {
       "baseline": "2.9.2",

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -170,7 +170,7 @@
     },
     "apr": {
       "baseline": "1.7.5",
-      "port-version": 1
+      "port-version": 2
     },
     "apr-util": {
       "baseline": "1.6.3",

